File: wrappers/javascript/src/_platform.ts
/**
 * Platform / runtime / SDK-version headers stamped on every control-plane
 * request. Mirrors the ``X-Stainless-*`` header family shipped by the
 * OpenAI and Anthropic SDKs; the ``X-Checkrd-SDK-*`` prefix keeps them
 * clearly namespaced to this project and makes them easy to filter in
 * ingestion logs.
 *
 * Why they matter:
 *   - An operator rolling out an SDK upgrade can see "we're seeing old
 *     (<0.3.0) clients still calling" from the dashboard, without asking
 *     every service owner to upgrade (Stripe / OpenAI pattern).
 *   - Runtime-specific bugs (e.g. "Cloudflare Workers pre-May-2026 miss
 *     traceparent") are identifiable without per-customer forensics.
 *   - Supply-chain incidents (a compromised transitive dep that flips the
 *     ``X-Checkrd-SDK-Lang`` to a wrong value) surface loudly.
 *
 * Detection is runtime-agnostic: the Node case reads ``process.versions``
 * / ``process.platform`` / ``process.arch``, other runtimes fall back to
 * ``"unknown"``. Never throws — the telemetry path must not gain a new
 * failure mode because of a platform-detection bug.
 */
import { VERSION } from "./_version.js";

/**
 * Runtime kinds we recognize explicitly. Anything else collapses to
 * ``"unknown"`` so the header is always a known finite set.
 */
export type RuntimeKind =
  | "node"
  | "bun"
  | "deno"
  | "workerd"
  | "edge-light"
  | "browser"
  | "unknown";

/**
 * Snapshot of platform detection, computed once per process and
 * memoized. The object returned by {@link platformHeaders} reads from
 * this snapshot — we do NOT recompute per request because nothing in
 * this module can change at runtime within a single process.
 */
export interface PlatformInfo {
  /** Language: always ``"javascript"`` for this SDK. */
  lang: "javascript";
  /** SDK package version (``VERSION`` constant). */
  sdkVersion: string;
  /** Runtime kind, see {@link RuntimeKind}. */
  runtime: RuntimeKind;
  /** Runtime version string (e.g. ``"20.11.1"``), or ``"unknown"``. */
  runtimeVersion: string;
  /** OS family: ``"linux"`` / ``"darwin"`` / ``"win32"`` / ``"unknown"``. */
  os: string;
  /** CPU architecture: ``"x64"`` / ``"arm64"`` / ``"unknown"``. */
  arch: string;
}

let _cached: PlatformInfo | null = null;

/**
 * Detect the current JS runtime. Cheap but runs on every import path;
 * we memoize via {@link platformInfo}.
 */
function detectRuntime(): { kind: RuntimeKind; version: string } {
  const g = globalThis as unknown as {
    Deno?: { version?: { deno?: string } };
    Bun?: { version?: string };
    EdgeRuntime?: unknown;
    navigator?: { userAgent?: unknown };
    process?: {
      versions?: { node?: string; bun?: string; workerd?: string };
    };
  };

  if (g.Deno?.version?.deno) {
    return { kind: "deno", version: g.Deno.version.deno };
  }
  if (g.Bun?.version) {
    return { kind: "bun", version: g.Bun.version };
  }
  if (g.process?.versions?.workerd) {
    return { kind: "workerd", version: g.process.versions.workerd };
  }
  if (g.EdgeRuntime !== undefined) {
    return { kind: "edge-light", version: "unknown" };
  }
  if (g.process?.versions?.node) {
    return { kind: "node", version: g.process.versions.node };
  }
  // A real browser is an acknowledged-dangerous deployment (see
  // `isRealBrowser` in index.ts) but still a valid runtime label —
  // operators may want to see these to decide whether to block.
  if (typeof g.navigator?.userAgent === "string") {
    return { kind: "browser", version: "unknown" };
  }
  return { kind: "unknown", version: "unknown" };
}

/**
 * Return the memoized {@link PlatformInfo}. Computed on first call,
 * reused thereafter. Safe on every runtime — never throws.
 */
export function platformInfo(): PlatformInfo {
  if (_cached !== null) return _cached;
  const rt = detectRuntime();
  const proc = (globalThis as unknown as {
    process?: { platform?: string; arch?: string };
  }).process;
  _cached = {
    lang: "javascript",
    sdkVersion: VERSION,
    runtime: rt.kind,
    runtimeVersion: rt.version,
    os: proc?.platform ?? "unknown",
    arch: proc?.arch ?? "unknown",
  };
  return _cached;
}

/**
 * Reset the memoized snapshot. Test-only — production code has no
 * reason to refresh platform detection mid-process.
 */
export function __resetPlatformInfoForTesting(): void {
  _cached = null;
}

/**
 * Return the platform headers that should be attached to every
 * control-plane request. All six are always set — a missing value is
 * sent as ``"unknown"`` rather than omitted so ingestion can
 * distinguish "old SDK that didn't send the header" from "newer SDK
 * that couldn't detect the field".
 *
 * Header names mirror ``X-Stainless-*`` semantics but live under
 * ``X-Checkrd-SDK-*`` to keep the namespace clean and greppable.
 */
export function platformHeaders(
  info: PlatformInfo = platformInfo(),
): Record<string, string> {
  return {
    "X-Checkrd-SDK-Lang": info.lang,
    "X-Checkrd-SDK-Version": info.sdkVersion,
    "X-Checkrd-SDK-Runtime": info.runtime,
    "X-Checkrd-SDK-Runtime-Version": info.runtimeVersion,
    "X-Checkrd-SDK-OS": info.os,
    "X-Checkrd-SDK-Arch": info.arch,
  };
}

File: wrappers/python/src/checkrd/_platform.py
"""Platform / runtime / SDK-version headers for control-plane requests.

Mirrors the JS SDK's ``_platform.ts`` and the ``X-Stainless-*`` header
family shipped by the OpenAI and Anthropic SDKs. The ``X-Checkrd-SDK-*``
prefix keeps the headers clearly namespaced so ingestion can grep them
apart from the underlying HTTP library's headers.

Why they matter:
  - An operator rolling out an SDK upgrade can see "we still have old
    (<0.3.0) clients calling" from the dashboard, without asking every
    service owner to upgrade (Stripe / OpenAI pattern).
  - Python-version-specific bugs (e.g. "3.9 users see this, 3.12 does
    not") surface without per-customer forensics.
  - Supply-chain incidents — a compromised transitive dep that flips
    ``X-Checkrd-SDK-Lang`` to a wrong value — are visible immediately.

Detection never raises. Any field we can't resolve sends as
``"unknown"`` rather than being omitted so ingestion can distinguish
"older SDK that didn't send the header" from "newer SDK that couldn't
detect the field".
"""

from __future__ import annotations

import platform
import sys
import uuid
from dataclasses import dataclass
from typing import Dict, Optional

from checkrd._version import __version__


def new_idempotency_key() -> str:
    """Generate a fresh ``Idempotency-Key`` value for a control-plane POST.

    Mirrors the JS SDK's ``newIdempotencyKey()``. Format: ``checkrd-<uuid4>``.

    **Generate once per logical operation, then reuse across retries.**
    The Stripe convention is the same key on every attempt of the same
    request so the control plane can dedupe — a fresh key per attempt
    defeats the entire point of idempotency. The :class:`TelemetryBatcher`
    and the public-key registrar both follow this pattern: one key
    captured outside the retry loop, reused on every attempt.
    """
    return f"checkrd-{uuid.uuid4()}"


@dataclass(frozen=True)
class PlatformInfo:
    """Snapshot of detected platform information.

    Computed once per process via :func:`platform_info` and memoized —
    nothing in this snapshot changes at runtime within a single process,
    so the telemetry hot path should not pay for re-detection.

    Attributes:
        lang: Language identifier — always ``"python"`` for this SDK.
        sdk_version: Package version (``__version__``).
        runtime: Python implementation (``"cpython"``, ``"pypy"``, …).
        runtime_version: Runtime version (e.g. ``"3.12.0"``).
        os: OS family (``"darwin"``, ``"linux"``, ``"windows"``).
        arch: CPU architecture (``"arm64"``, ``"x86_64"``, …).
    """

    lang: str
    sdk_version: str
    runtime: str
    runtime_version: str
    os: str
    arch: str


_cached: Optional[PlatformInfo] = None


def _detect() -> PlatformInfo:
    """Detect the current Python runtime. Cheap; never raises."""
    try:
        impl = platform.python_implementation().lower()
    except Exception:
        impl = "unknown"
    try:
        impl_version = platform.python_version()
    except Exception:
        impl_version = "unknown"
    try:
        # `sys.platform` is canonical for cross-ref with CI matrix names.
        os_name = sys.platform.lower() or "unknown"
    except Exception:
        os_name = "unknown"
    try:
        arch = platform.machine().lower() or "unknown"
    except Exception:
        arch = "unknown"
    return PlatformInfo(
        lang="python",
        sdk_version=__version__,
        runtime=impl,
        runtime_version=impl_version,
        os=os_name,
        arch=arch,
    )


def platform_info() -> PlatformInfo:
    """Return the memoized :class:`PlatformInfo` snapshot."""
    global _cached  # noqa: PLW0603
    if _cached is None:
        _cached = _detect()
    return _cached


def _reset_platform_info_for_testing() -> None:
    """Reset the memoized snapshot. Test-only hook."""
    global _cached  # noqa: PLW0603
    _cached = None


def platform_headers(info: Optional[PlatformInfo] = None) -> Dict[str, str]:
    """Return the ``X-Checkrd-SDK-*`` header set for a control-plane request.

    All six headers are always included. Missing fields are sent as
    ``"unknown"`` so ingestion can distinguish the older-SDK case from
    the can't-detect case.

    Args:
        info: Optional override — primarily for testing. Defaults to
            :func:`platform_info`.
    """
    if info is None:
        info = platform_info()
    return {
        "X-Checkrd-SDK-Lang": info.lang,
        "X-Checkrd-SDK-Version": info.sdk_version,
        "X-Checkrd-SDK-Runtime": info.runtime,
        "X-Checkrd-SDK-Runtime-Version": info.runtime_version,
        "X-Checkrd-SDK-OS": info.os,
        "X-Checkrd-SDK-Arch": info.arch,
    }


def default_control_headers(
    api_key: str,
    *,
    api_version: str = "",
    idempotency_key: Optional[str] = None,
    content_type: str = "application/json",
) -> Dict[str, str]:
    """Standard header set for a Checkrd control-plane request.

    Consolidated so the telemetry batcher, public-key registrar, and
    SSE receiver all send identical metadata — operators looking at
    ingestion logs see one consistent shape.

    Always-on:
        - ``Content-Type`` (default ``application/json``; pass ``""`` to omit)
        - ``X-API-Key``
        - ``User-Agent: checkrd-python/<version>``
        - ``X-Checkrd-SDK-*`` family

    Optional:
        - ``Checkrd-Version`` when ``api_version`` is a non-empty string
          (Stripe-style date pin).
        - ``Idempotency-Key`` when ``idempotency_key`` is provided —
          callers generate and reuse a UUID per retry loop, so this
          helper does NOT auto-generate one (unlike the JS side, which
          does fresh-per-call because the JS retry loop captures the
          header set once up front).
    """
    headers: Dict[str, str] = {
        "X-API-Key": api_key,
        "User-Agent": f"checkrd-python/{__version__}",
    }
    if content_type:
        headers["Content-Type"] = content_type
    headers.update(platform_headers())
    if api_version:
        headers["Checkrd-Version"] = api_version
    if idempotency_key is not None:
        headers["Idempotency-Key"] = idempotency_key
    return headers
